xdsl/backend/riscv/lowering/convert_memref_to_riscv.py

```diff
@@ -141,7 +145,74 @@ def memref_shape_ops(
                 comment="multiply by element size",
             ),
             ptr := riscv.AddOp(
-                mem, offset_bytes, rd=riscv.IntRegisterType.unallocated()
+                src_ptr, offset_bytes, rd=riscv.IntRegisterType.unallocated()
+            ),
+        ]
+    )
+
+    return ops, ptr.rd
+
+
+def pointer_offsets_ops(
+    src_ptr: SSAValue,
+    pairs: Iterable[tuple[SSAValue, int]],
+    element_type: Attribute,
+) -> tuple[list[Operation], SSAValue]:
+    """
+    Returns an ssa value representing a pointer into the memref at the given indices.
+    'pairs' consists of the indices as SSA values and the corresponding stride of the dimension being indexed in
+    number of elements. 'element_type' must be the element type of the memref being indexed.
+    """
+    bitwidth = bitwidth_of_type(element_type)
+    if bitwidth % 8:
+        raise DiagnosticException(
+            f"Cannot create offset for element type {element_type}"
+            f" with bitwidth {bitwidth}"
+        )
+    bytes_per_element = bitwidth // 8
+
+    pairs = tuple((index, factor) for index, factor in pairs if factor != 0)
+
+    if not pairs:
+        # Scalar memref
+        return ([], src_ptr)
+
+    ops: list[Operation] = []
+    res_ptr = src_ptr
+
+    for index, factor in pairs:
+        if factor == 1:
+            ops.append(
+                head_op := riscv.AddOp(
+                    res_ptr, index, rd=riscv.IntRegisterType.unallocated()
+                )
+            )
+            res_ptr = head_op.rd
+        else:
+            ops.extend(
+                (
+                    factor_op := riscv.LiOp(factor),
+                    offset_op := riscv.MulOp(
+                        index, factor_op.rd, rd=riscv.IntRegisterType.unallocated()
+                    ),
+                    head_op := riscv.AddOp(
+                        res_ptr, offset_op.rd, rd=riscv.IntRegisterType.unallocated()
+                    ),
+                )
+            )
+        res_ptr = head_op.rd
+
+    ops.extend(
+        [
+            bytes_per_element_op := riscv.LiOp(bytes_per_element),
+            offset_bytes := riscv.MulOp(
+                res_ptr,
+                bytes_per_element_op.rd,
+                rd=riscv.IntRegisterType.unallocated(),
+                comment="multiply by element size",
+            ),
+            ptr := riscv.AddOp(
+                src_ptr, offset_bytes, rd=riscv.IntRegisterType.unallocated()
```

Comment:
The code here seems problematic. Specifically, in the `for` before we keep on adding onto `res_ptr` and then we are multiplying the `res_ptr` with `bytes_per_element`. 

That means we are not multiplying the indices but rather the pointer value. `offset_bytes` is also misleadin given its actually the pointer/memref

Reply:
Yep, I think I got a little lost in the sauce here, the new version should be correct.